### PR TITLE
Change reports framework

### LINKED ISSUE #1920: Restore column selection functionality for reports view
Argo's report view displays a list of columns. It used to be possible for users to interactively select specific columns to display in this view. This ticket is the restore that functionality.

**Additional background**

In fall 2019, column selection functionality ceased to work following a set of javascript upgrades made during the fall maintenance workcycle. As I understand it, some of the libraries the old functionality relied on are no longer maintained, and keeping the function in Argo as it was presented a security risk.

The ability to select columns for downloaded reports was restored last fall, but restoring in-browser column selection had to be deferred into the next maintenance work cycle owing to time and resourcing constraints.

### LINKED ISSUE #2124: Implement reports view with new framework from #4147
With the framework chosen in #4147, re-implement the reports view.

Step 1:  think of useful tests -- perhaps feature tests taking the reports through its paces.
Step 2:  re-implement the reports view.
Step 3:  if you have enough for a PR now, make a PR and get it reviewed by Andrew.

Some additional things we'd like in the implementation:
NOTE: it's fine for these to be tackled as separate tickets / PRs ... it's up to the implementer of THIS ticket
- column selection functionality in browser reports view #1920
- list of fields allowed for reports #4077
- druids are links to the objects #923

### LINKED ISSUE #4147: analysis:  determine framework candidate(s) for reports view
The Argo reports view uses a javascript library that is no longer supported. This limits our ability to improve the view, as seen in https://github.com/sul-dlss/argo/issues/1920. 

This ticket is to investigate new framework(s) that meet our requirements that we can improve and maintain.

There is another ticket (#2124) to implement reports using the new framework chosen;  if you already did that as part of this investigation, great!

### What are our behavior requirements?

I suggest you play around with the reports view in Argo to understand what it does now.

@astridu has a design proposal for what we're aiming for in an improved report view:
https://projects.invisionapp.com/share/U613AO7XYT5B#/screens/474040067_report_Select_Columns

- What you can do with reports now in Argo:
  - Load rows progressively ("pagination"), not all at once
  - Render scrollbar
  - Download rows as CSV
  - Select which columns are included in the download
- Plus these new (or restored) features:
  - #1920
  - #923
  - #4077
- Optional ("nice to have"):
  - Re-order columns
  - Sort by clicking headers

I also suggest you check with Andrew again to make sure your list of requirements includes all of his requirements.

### What are the technical requirements?

It doesn't absolutely have to be javascript, but that seems likely based on UI interaction.  Maybe someone has something clever in ruby and/or with stimulus?

5 seconds of searching (the person who takes this ticket should definitely do more) got me to https://tabulator.info/.

### More background

This image is from 2020, in ticket #1920:

image

### PR DESCRIPTION
**Pull request by mjgiarlo**
# TODO

- [x] There’s no indication that anything happened when I clicked “Download CSV”. The CSV came back surprisingly fast but I clicked multiple times while waiting and got multiple downloads
- [x] ~~It would be nice to have a visual indication when more rows are going to be loaded. Loading is so smooth that it’s almost hard to tell it’s happening.~~
- [x] I wonder if we should go without the sort arrows. They appear to sort only what the client has loaded (which makes sense) but that can lead people to believe the results are complete before they are.

# Why was this change made?

Fixes #4147
Fixes #1920 
Fixes #2124 

* Stop installing foreman when already installed
* Tear out jqGrid and most jquery vestiges
* Use tabulator to build reports UI

# How was this change tested?

- [x] CI
- [x] QA/stage

# What's the accessibility impact?

* aXe
  * From 34 issues to 3 issues
* WAVE
  * From 30 errors & 1,177 alerts to 7 errors and 5 alerts
* Siteimprove
  * From 6 issues (202+ instances) to 3 issues (5 instances)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -217,7 +217,6 @@ Argo uses Blacklight and ActiveFedora to expose the repository contents, and `do
 - RSolr
 - Sidekiq
 - Bootstrap
-- JQuery
```
Comment (mjgiarlo) [inline README.md:220]: JQuery is no longer a key component of Argo. (Argo still depends on it, but only because Blacklight 7 does.)

File: app/assets/stylesheets/argo.scss [hunk 2]
```diff
@@ -59,10 +58,6 @@ legend {
   color: red;
 }
 
-#report_container {
-  margin-top: 1.5em;
-}
```
Comment (mjgiarlo) [inline app/assets/stylesheets/argo.scss:64]: Only used in old reports UI.

File: app/assets/stylesheets/application.scss [hunk 3]
```diff
@@ -1,15 +1,14 @@
...
 @import 'bootstrap-overrides';
 // Override before importing Blacklight
 @import 'blacklight-frontend/app/assets/stylesheets/blacklight/blacklight';
 @import 'blacklight-hierarchy/app/assets/stylesheets/blacklight/hierarchy/hierarchy';
+@import 'simple-datatables/dist/style';
 
 @import "argo";
 @import "bulk-actions";
 @import "errors";
-@import "dialog";
```
Comment (mjgiarlo) [inline app/assets/stylesheets/application.scss:12]: Only used in old reports UI.

File: app/assets/stylesheets/jquery-ui-1.12.1.custom/jquery-ui.theme.css [hunk 4]
```diff
@@ -1,444 +0,0 @@
-/*!
```
Comment (mjgiarlo) [inline app/assets/stylesheets/jquery-ui-1.12.1.custom/jquery-ui.theme.css:1]: Only used in old reports UI.

File: app/assets/stylesheets/registration.scss [hunk 5]
```diff
@@ -1,53 +1,3 @@
...
-	background: #fff;
-	color: black;
-}
-
-.ui-state-active,
-.ui-widget-content .ui-state-active {
-	border: 1px solid $primary;
-	background: $primary;
-	font-weight: 400;
-	color: #fff;
```
Comment (mjgiarlo) [inline app/assets/stylesheets/registration.scss:52]: Only used in old reports UI.

File: app/assets/stylesheets/registration/registration.scss [hunk 6]
```diff
@@ -1,35 +0,0 @@
-#t_data {
```
Comment (mjgiarlo) [inline app/assets/stylesheets/registration/registration.scss:1]: Only used in old reports UI.

File: app/assets/stylesheets/registration/ui-override.scss [hunk 7]
```diff
@@ -1,30 +0,0 @@
-.edit-cell input {
```
Comment (mjgiarlo) [inline app/assets/stylesheets/registration/ui-override.scss:1]: Only used in old reports UI.

File: app/controllers/report_controller.rb [hunk 8]
```diff
@@ -10,7 +10,7 @@ def rsolr_request_error(exception)
   end
 
   def data
-    params[:sord] ||= "asc"
+    params[:sort] ||= "asc"
```
Comment (mjgiarlo) [inline app/controllers/report_controller.rb:13]: Typo. Unrelated to PR 🤷🏻‍♂️

File: app/controllers/report_controller.rb [hunk 9]
```diff
@@ -31,16 +31,6 @@ def data
...
 
-  def download
-    fields = params["fields"].present? ? params.delete("fields").split(/\s*,\s*/) : nil
-    params[:per_page] = 10
-    params[:page] = 1
-    response.headers["Content-Type"] = "application/octet-stream"
-    response.headers["Content-Disposition"] = "attachment; filename=report.csv"
-    response.headers["Last-Modified"] = Time.now.utc.rfc2822 # HTTP requires GMT date/time
-    self.response_body = Report.new(params, fields, current_user:).to_csv
-  end
```
Comment (mjgiarlo) [inline app/controllers/report_controller.rb:42]: No longer needed, since the simple-datatables library will export to CSV on the client-side for us. (Would be good for @andrewjbtw to take a gander at the CSV to make sure it looks/works as expected.)

File: app/javascript/argo.js [hunk 10]
```diff
@@ -19,11 +16,6 @@ export default class Argo {
       Blacklight.activate()
     }
 
-    report() {
-      if (document.querySelector('[data-controller="report"]'))
-        initializeReport()
-    }
```
Comment (mjgiarlo) [inline app/javascript/argo.js:25]: No longer needs special handling---this is now done via Stimulus.

File: app/javascript/controllers/object_reporter.js [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+  connect() {
+    if (this.table !== undefined) return; // Prevent negative side effects if hook is called more than once per page load
+
+    this.table = new DataTable("#objectsTable", {
+      columns: this.transformColumns(),
+      searchable: true,
+      sortable: true,
+      scrollY: "70vh",
+      paging: false,
```
Comment (mjgiarlo) [inline app/javascript/controllers/object_reporter.js:20]: Approximate the former look & feel.

File: app/javascript/controllers/object_reporter.js [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+      rowNavigation: true,
+      tabIndex: 1,
```
Comment (mjgiarlo) [inline app/javascript/controllers/object_reporter.js:22]: Accessibility improvements.

File: app/javascript/modules/report.js [hunk 13]
```diff
@@ -1,82 +0,0 @@
-/*
```
Comment (mjgiarlo) [inline app/javascript/modules/report.js:1]: Byebye, old report.

File: app/models/report.rb [hunk 14]
```diff
@@ -14,6 +14,8 @@ class << self
     include ValueHelper
   end
 
+  COLUMN_SELECTOR_COLUMN_SIZE = 3
```
Comment (mjgiarlo) [inline app/models/report.rb:17]: Helps display report columns in neatly lined up columns.

File: app/models/report.rb [hunk 15]
```diff
@@ -151,13 +153,8 @@ class << self
...
   ].freeze
 
-  COLUMN_MODEL = REPORT_FIELDS.collect do |spec|
+  COLUMN_MODEL = REPORT_FIELDS.map do |spec|
     {
-      "name" => spec[:field],
-      "jsonmap" => spec[:field],
-      "label" => spec[:label],
-      "index" => spec[:field],
-      "width" => spec[:width],
```
Comment (mjgiarlo) [inline app/models/report.rb:160]: No longer used.

File: app/views/layouts/blacklight/base.html.erb [hunk 16]
```diff
@@ -19,8 +19,8 @@
   </head>
   <body class="<%= render_body_class %>">
     <nav id="skip-link" role="navigation" aria-label="<%= t('blacklight.skip_links.label') %>">
-      <%= link_to t('blacklight.skip_links.search_field'), '#search_field', class: 'element-invisible element-focusable rounded-bottom py-2 px-3', data: { turbolinks: 'false' } %>
       <%= link_to t('blacklight.skip_links.main_content'), '#main-container', class: 'element-invisible element-focusable rounded-bottom py-2 px-3', data: { turbolinks: 'false' } %>
```
Comment (mjgiarlo) [inline app/views/layouts/blacklight/base.html.erb:22]: Skip to main link should be first link for a11y.

File: app/views/report/_column_selector.html.erb [hunk 17]
```diff
@@ -1,6 +0,0 @@
-<div class="d-none" id="column_selector">
```
Comment (mjgiarlo) [inline app/views/report/_column_selector.html.erb:1]: Inlined this in _document_list.html.erb because it's only used there.

File: bin/dev [hunk 18]
```diff
@@ -1,9 +1,8 @@
...
+#!/usr/bin/env bash
 
-if ! foreman version &> /dev/null
-then
-    echo "Installing foreman..."
-    gem install foreman
+if ! gem list foreman -i --silent; then
+  echo "Installing foreman..."
+  gem install foreman
 fi
```
Comment (mjgiarlo) [inline bin/dev:6]: Prior formulation is the outdated cssbundling way of doing things (which installs foreman on every single invocation of `bin/dev`). Updated to latest way because less annoying.

File: config/routes.rb [hunk 19]
```diff
@@ -76,7 +76,6 @@
 
   match "report", to: "report#index", via: %i[get post], as: "report"
   match "report/data", to: "report#data", via: %i[get post], as: "report_data"
-  match "report/download", to: "report#download", via: %i[get post], as: "report_download"
```
Comment (mjgiarlo) [inline config/routes.rb:79]: No longer needed.

File: app/views/report/_document_list.html.erb [hunk 20]
```diff
@@ -1,16 +1,49 @@
...
-      total_rows: <%= @response['response']['numFound'] %>,
-      data_url: <%= report_data_url(params.to_unsafe_h.merge(format: 'json')).to_json.html_safe %>,
-      download_url: <%= report_download_url(params.to_unsafe_h).to_json.html_safe %>,
-      column_model: <%= Report::COLUMN_MODEL.to_json.html_safe %>
-    }
-  </script>
-<% end %>
+<div class="mt-3"
+     data-controller="object-reporter"
+     data-object-reporter-data-url-value=<%= report_data_url(params.to_unsafe_h.merge(format: 'json', rows: 99_999_999)) %>
```
Comment (mjgiarlo) [inline app/views/report/_document_list.html.erb:3]: This was significantly easier than trying to page through Solr results dependent on how far down the user has scrolled.

So, we'll want to test reports with a large number of results to see how badly performance suffers. Cc: @andrewjbtw

File: app/controllers/report_controller.rb (hunk 8 above)
Comment (ndushay) [reply app/controllers/report_controller.rb:13]: geez.  It should have been 'sword'.

File: app/javascript/controllers/object_reporter.js (hunk 11 above)
Comment (ndushay) [reply app/javascript/controllers/object_reporter.js:20]: is there a design by Astrid or other UI guidance?

File: app/javascript/controllers/object_reporter.js [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+      labels: {
+        noRows: "Loading report entries...",
+      },
+    });
+
+    // This removes the pagination list adds an aria label for its parent nav,
```
Comment (ndushay) [inline app/javascript/controllers/object_reporter.js:28]: garbled language in comment.

File: app/views/report/_document_list.html.erb (hunk 20 above)
Comment (ndushay) [reply app/views/report/_document_list.html.erb:3]: if performance is poor, i wonder if we can send a Solr query that only returns the potential fields of interest and avoids large fields of no interest.

File: app/javascript/application.js [hunk 22]
```diff
@@ -1,18 +1,17 @@
 // Entry point for the build script in your package.json
 
-import 'jquery'
+import 'jquery' // Blacklight 7 needs jQuery. Remove when we upgrade to Blacklight 8?
 import bootstrap from  "bootstrap/dist/js/bootstrap"
 global.bootstrap = bootstrap // Required for Blacklight 7 so it can manage the modals
 
 import 'blacklight-frontend/app/assets/javascripts/blacklight/blacklight'
+import 'simple-datatables'
```
Comment (ndushay) [inline app/javascript/application.js:8]: comment that this is for reports view??  I don't feel strongly.

**Review by ndushay — COMMENTED** [review]
lurv that it's so much less code.

Wondering if the report_view_spec or the report_spec have been thought through to ponder if  additional tests would be helpful to our future selves?

Not approving only because I'm wondering how stable this code will be as Andrew may need some changes.

File: app/javascript/controllers/object_reporter.js (hunk 11 above)
Comment (mjgiarlo) [reply app/javascript/controllers/object_reporter.js:20]: There is an old design that pre-dates Astrid, I believe, but for the current PR I've aimed for "as close to parity as we can get using standard simple-datatables config"

File: app/javascript/application.js (hunk 22 above)
Comment (mjgiarlo) [reply app/javascript/application.js:8]: Added.

File: app/javascript/controllers/object_reporter.js (hunk 21 above)
Comment (mjgiarlo) [reply app/javascript/controllers/object_reporter.js:28]: Copypasta, will improve and push.

File: app/views/report/_document_list.html.erb (hunk 20 above)
Comment (mjgiarlo) [reply app/views/report/_document_list.html.erb:3]: Yeah, that should certainly go on the list of changes to consider.

**Comment by mjgiarlo** [comment]
@ndushay 💬 
> Wondering if the report_view_spec or the report_spec have been thought through to ponder if additional tests would be helpful to our future selves?

I opted not to add testing since it's primarily a UI-related change. FWIW, the report data is coming from the `Report` model which is tested fairly thoroughly in `spec/models/report_spec.rb`.
 
> Not approving only because I'm wondering how stable this code will be as Andrew may need some changes.

🆗

**Comment by mjgiarlo** [comment]
BUG: This only loads 99 rows from the data.

**Comment by andrewjbtw** [comment]
The headers and the rows with data do not line up correctly.

Screenshot 2023-09-14 at 2 28 41 PM

In the screenshot above, the alignment really starts to go off after source id. Exactly where the alignment goes off seems to depend on the data that's retrieved.

File: app/controllers/report_controller.rb [hunk 23]
```diff
@@ -5,42 +5,29 @@ class ReportController < CatalogController
...
 
     respond_to do |format|
       format.json do
         render json: {
           page: params[:page].to_i,
-          records: @report.num_found,
-          total: (@report.num_found / rows_per_page.to_f).ceil,
-          rows: @report.report_data
+          last_page: (@report.num_found / ROWS_PER_PAGE.to_f).ceil,
+          data: @report.report_data
```
Comment (mjgiarlo) [inline app/controllers/report_controller.rb:24]: Adjust what's returned to be what the client-side table library (Tabulator) wants.

File: app/controllers/report_controller.rb [hunk 24]
```diff
@@ -5,42 +5,29 @@ class ReportController < CatalogController
...
         }
       end
       format.xml { render xml: @report.report_data }
     end
   end
 
-  def content_types
-  end
```
Comment (mjgiarlo) [inline app/controllers/report_controller.rb:32]: Unused/cruft.

File: app/javascript/application.js [hunk 25]
```diff
@@ -1,18 +1,16 @@
...
-  await import('https://cdnjs.cloudflare.com/ajax/libs/free-jqgrid/4.15.5/jquery.jqgrid.src.js')
-
-  // Start argo after free-jqgrid has been loaded
-  new Argo().initialize()
-})
-import '@hotwired/turbo-rails'
+  // Start argo after Turbo has been loaded
+  new Argo().initialize();
+});
+import "@hotwired/turbo-rails";
```
Comment (mjgiarlo) [inline app/javascript/application.js:16]: My editor did some normalization here. I intend to follow up later with more JS normalization and perhaps linting in Argo. I can roll these mechanical changes back if anyone cares.

File: app/controllers/report_controller.rb [hunk 26]
```diff
@@ -10,37 +10,22 @@ def rsolr_request_error(exception)
   end
 
   def data
-    params[:sord] ||= "asc"
-    rows_per_page = params[:rows] ? params.delete(:rows).to_i : 10
-    params[:per_page] = rows_per_page * [params.delete(:npage).to_i, 1].max
+    params[:per_page] = params[:per_page] ? params[:per_page].to_i : Report::ROWS_PER_PAGE
```
Comment (mjgiarlo) [inline app/controllers/report_controller.rb:13]: Keeping this param around allows us to tweak the default rows-per-page value of 100.

File: app/controllers/report_controller.rb [hunk 27]
```diff
@@ -10,37 +10,22 @@ def rsolr_request_error(exception)
...
+    params[:page] ||= 1
```
Comment (mjgiarlo) [inline app/controllers/report_controller.rb:14]: The client-side tooling will automagically increment this param as it progressively loads report results via the server-side API (`ReportController` & `Report` model).

File: app/controllers/report_controller.rb [hunk 28]
```diff
@@ -10,37 +10,22 @@ def rsolr_request_error(exception)
...
 
     respond_to do |format|
       format.json do
         render json: {
           page: params[:page].to_i,
-          records: @report.num_found,
-          total: (@report.num_found / rows_per_page.to_f).ceil,
-          rows: @report.report_data
+          last_page: (@report.num_found / params[:per_page].to_f).ceil,
+          data: @report.report_data
```
Comment (mjgiarlo) [inline app/controllers/report_controller.rb:22]: These changes align the server-side impl with what the client-side wants.

File: app/javascript/controllers/object_reporter.js [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+    this.table = new Tabulator("#objectsTable", {
+      layout: "fitDataTable",
+      height: "70vh",
+      addRowPos: "bottom", //when adding a new row, add it to the bottom of the table
+      movableColumns: true, //allow column order to be changed
+      progressiveLoad: "scroll", //enable progressive loading
+      ajaxURL: this.dataUrlValue,
+      ajaxParams: this.dataUrlParamsValue,
+      columns: this.columnModelValue,
+    });
```
Comment (mjgiarlo) [inline app/javascript/controllers/object_reporter.js:24]: Look how little code we now need on the client side 😍

File: app/models/concerns/csv_concern.rb [hunk 30]
```diff
@@ -1,24 +0,0 @@
...
-      until @response.documents.empty?
-        report_data.each do |record|
-          csv << @fields.map { |f| record[f[:field]].to_s }
-        end
-        @params[:page] += 1
-        (@response,) = search_results(params)
-      end
-    end
-  end
-end
```
Comment (mjgiarlo) [inline app/models/concerns/csv_concern.rb:24]: No longer needed because the client-side library handles the CSV export for us.

File: app/models/report.rb [hunk 31]
```diff
@@ -6,32 +6,34 @@
...
 
+  COLUMN_SELECTOR_COLUMN_SIZE = 3
+  ROWS_PER_PAGE = 100
+
   REPORT_FIELDS = [
     {
       field: :druid, label: "Druid",
       proc: ->(doc) { doc.druid },
-      sort: true, default: true, width: 100, download_default: true
+      sort: true, default: true, width: 100
```
Comment (mjgiarlo) [inline app/models/report.rb:23]: We no longer need to distinguish between whether to make a field downloadable by default and whether to make a field visible by default. These concerns are now aligned. We only made this change as a concession when discovering how broken our outdated client-side library had become, I believe.

File: app/models/report.rb [hunk 32]
```diff
@@ -42,124 +44,122 @@ class << self
...
-      "label" => spec[:label],
-      "index" => spec[:field],
-      "width" => spec[:width],
-      "sortable" => spec[:sort],
-      "hidden" => (!spec[:default])
+      "field" => spec[:field],
+      "title" => spec[:label],
+      "visible" => spec[:default],
+      "minWidth" => spec[:width],
+      "headerSort" => spec[:sort]
```
Comment (mjgiarlo) [inline app/models/report.rb:162]: Align the returned hashes w/ what the client-side library expects.

File: app/models/report.rb [hunk 33]
```diff
@@ -170,31 +170,25 @@ class << self
...
   # @param [Array<String>] fields
-  def initialize(params = {}, fields = nil, current_user: NullUser.new)
+  def initialize(params = {}, current_user: NullUser.new)
     @current_user = current_user
-    @fields = if fields.nil?
-      REPORT_FIELDS
-    else
-      REPORT_FIELDS.select { |f| fields.include?(f[:field].to_s) }
-        .sort { |a, b| fields.index(a[:field].to_s) <=> fields.index(b[:field].to_s) }
-    end
```
Comment (mjgiarlo) [inline app/models/report.rb:188]: We turn fields on and off via Stimulus now.

File: app/models/report.rb [hunk 34]
```diff
@@ -170,31 +170,25 @@ class << self
...
+    @fields = REPORT_FIELDS
     @params = params
-    @params[:page] ||= 1
```
Comment (mjgiarlo) [inline app/models/report.rb:190]: Set in `ReportController` now.

File: app/models/report.rb [hunk 35]
```diff
@@ -170,31 +170,25 @@ class << self
...
     (@response,) = search_results(@params)
     @num_found = @response["response"]["numFound"].to_i
   end
 
   def druids(opts = {})
     params[:page] = 1
-    params[:per_page] = 100
+    params[:per_page] = ROWS_PER_PAGE
```
Comment (mjgiarlo) [inline app/models/report.rb:191]: Don't repeat the magic number.

File: app/models/report.rb [hunk 36]
```diff
@@ -246,9 +240,8 @@ def search_service_class
   # @param [Array<Hash>] fields
   # @return [Array<Hash(Symbol => String)>]
   def docs_to_records(docs, fields = REPORT_FIELDS)
-    result = []
-    docs.each_with_index do |doc, index|
-      row = fields.to_h do |spec|
```
Comment (mjgiarlo) [inline app/models/report.rb:251]: Make better use of enumerable methods.

File: app/controllers/report_controller.rb [hunk 37]
```diff
@@ -10,35 +10,28 @@ def rsolr_request_error(exception)
...
   end
 
-  def content_types
-  end
-
   def download
-    fields = params["fields"].present? ? params.delete("fields").split(/\s*,\s*/) : nil
-    params[:per_page] = 10
-    params[:page] = 1
+    params[:rows] = 99_999_999
```
Comment (mjgiarlo) [inline app/controllers/report_controller.rb:30]: Get the whole result set in the downloaded CSV.

File: spec/models/report_spec.rb [hunk 38]
```diff
@@ -12,7 +12,7 @@
     solr_conn.commit
   end
 
-  context "csv" do
+  describe "#to_csv" do
```
Comment (mjgiarlo) [inline spec/models/report_spec.rb:15]: More idiomatic RSpec usage.

File: app/javascript/controllers/object_reporter.js [hunk 39]
```diff
@@ -0,0 +1,83 @@
...
+
+  downloadCSV() {
+    const flashContainer = document.getElementsByClassName("flash_messages")[0];
+    const downloadInProgressFlash = `
+      <div id="csv-download-flash" class="alert alert-info">
+        Your CSV report is being generated... Please hold.
+        <a class="btn-close close" data-dismiss="alert" data-bs-dismiss="alert" href="#">&times;</a>
+      </div>
+    `;
+    flashContainer.innerHTML = downloadInProgressFlash;
```
Comment (mjgiarlo) [inline app/javascript/controllers/object_reporter.js:37]: Pop up, and later clean up, a flash message so users know the CSV may take a little bit to generate and download (for especially large data sets)

File: app/controllers/report_controller.rb [hunk 40]
```diff
@@ -10,35 +10,33 @@ def rsolr_request_error(exception)
...
-  def content_types
-  end
-
   def download
-    fields = params["fields"].present? ? params.delete("fields").split(/\s*,\s*/) : nil
-    params[:per_page] = 10
-    params[:page] = 1
-    response.headers["Content-Type"] = "application/octet-stream"
-    response.headers["Content-Disposition"] = "attachment; filename=report.csv"
+    response.headers.delete("Content-Length")
```
Comment (justinlittman) [inline app/controllers/report_controller.rb:30]: Why is this necessary?

File: app/controllers/report_controller.rb [hunk 41]
```diff
@@ -10,35 +10,33 @@ def rsolr_request_error(exception)
...
+    response.headers["Cache-Control"] = "no-cache"
+    response.headers["Content-Type"] = "text/csv"
+    response.headers["Content-Disposition"] = "attachment; filename=\"report.csv\""
     response.headers["Last-Modified"] = Time.now.utc.rfc2822 # HTTP requires GMT date/time
-    self.response_body = Report.new(params, fields, current_user:).to_csv
+    response.headers["X-Accel-Buffering"] = "no"
+
+    response.status = 200
```
Comment (justinlittman) [inline app/controllers/report_controller.rb:37]: Is this necessary?

**Review by justinlittman — APPROVED** [review]

File: app/controllers/report_controller.rb (hunk 40 above)
Comment (mjgiarlo) [reply app/controllers/report_controller.rb:30]: @justinlittman This is used to let Rack know to stream the response instead of filling a buffer. Learned about this technique by doing research about how to stream large CSV results to the user, e.g., https://medium.com/reflektive-engineering/faster-csv-downloads-using-enumerator-7e9b94b870d3.

File: app/controllers/report_controller.rb (hunk 41 above)
Comment (mjgiarlo) [reply app/controllers/report_controller.rb:37]: Come to think of it, this may be cruft. Will test and remove if so.

**Review by aaron-collier — APPROVED** [review]

File: app/controllers/report_controller.rb (hunk 27 above)
Comment (ndushay) [reply app/controllers/report_controller.rb:14]: this would make a good comment in the code.

File: app/controllers/report_controller.rb (hunk 28 above)
Comment (ndushay) [reply app/controllers/report_controller.rb:22]: ditto with code comment.

File: app/models/report.rb (hunk 14 above)
Comment (ndushay) [reply app/models/report.rb:17]: another spot where a comment in the code would be useful?

File: app/models/report.rb [hunk 42]
```diff
@@ -42,124 +45,129 @@ class << self
...
+      "visible" => spec[:default],
+      "minWidth" => spec[:width],
+      # Disable sorting due to behavior that will confuse users, namely the
+      # likelihood of a user attempting to sort a column before the full report
+      # data set has loaded, namely that we are loading data sets progressively
+      # as users scroll down the reports view to keep performance satisfactory.
+      # Sorting a partial data set only to scroll to the bottom and have more
+      # rows load will look strange, and give users false confident that they
+      # are looking at the complete data set.
+      "headerSort" => false # spec[:sort]
```
Comment (ndushay) [inline app/models/report.rb:170]: not allowing sorting makes me sad.  :-(

File: app/models/report.rb [hunk 43]
```diff
@@ -1,37 +1,40 @@
...
       proc: ->(doc) { doc.druid },
-      sort: true, default: true, width: 100, download_default: true
+      sort: true, default: true, width: 100
     },
     {
       field: :purl, label: "Purl",
       proc: ->(doc) { File.join(Settings.purl_url, doc.druid) },
       solr_fields: %w[id],
-      sort: false, default: false, width: 100, download_default: true
+      sort: false, default: false, width: 100
```
Comment (ndushay) [inline app/models/report.rb:30]: is it really the case that all field widths should be equal?  Or maybe the user can adjust this once it loads?

File: package.json [hunk 44]
```diff
@@ -8,11 +8,11 @@
     "blacklight-hierarchy": "^6.0.2",
     "bootstrap": "^5.1.0",
     "esbuild": "^0.12.28",
-    "free-jqgrid": "^4.15.5",
-    "jquery": "^3.4.1",
```
Comment (ndushay) [inline package.json:12]: whoop whoop!

File: app/models/report.rb [hunk 45]
```diff
@@ -222,6 +233,25 @@ def report_data
     docs_to_records(@response.documents)
   end
 
+  ##
+  # Converts the `report_data` into CSV data
+  #
+  # @return [Enumerator] data in CSV format
+  def to_csv
```
Comment (ndushay) [inline app/models/report.rb:240]: wouldn't "csv" be more idiomatic ruby?

**Review by ndushay — COMMENTED** [review]
a few post merge comments for your consideration.

File: app/models/report.rb (hunk 42 above)
Comment (mjgiarlo) [reply app/models/report.rb:170]: It's a bummer, but Andrew requested it for the reason listed, and I can't blame him! It would be very confusing. The old UI didn't allow sorting either so nothing's lost, at least.

File: app/controllers/report_controller.rb (hunk 27 above)
Comment (mjgiarlo) [reply app/controllers/report_controller.rb:14]: Added.

File: app/controllers/report_controller.rb (hunk 28 above)
Comment (mjgiarlo) [reply app/controllers/report_controller.rb:22]: Done.

File: app/models/report.rb (hunk 14 above)
Comment (mjgiarlo) [reply app/models/report.rb:17]: Done.

File: app/models/report.rb (hunk 43 above)
Comment (mjgiarlo) [reply app/models/report.rb:30]: All field widths are not equal, but most of them are set to `100`. These are effectively minimum widths which we can suss out via line 162.

File: app/models/report.rb (hunk 45 above)
Comment (mjgiarlo) [reply app/models/report.rb:240]: I don't think so, no.`to_{format}` methods are quite common, e.g., `to_json`, `to_h`, `to_a`, `to_s`,
